Warning on unintended implicit drops

jashug opened the pull request:
Rendered: https://github.com/jashug/rfcs/blob/scope_drop/text/0000-implicit-drop-warning.md

thomcc commented:
This will cause significant churn across nearly every rust library, complicating nearly every generic function definition, and being something every new Rust user needs to understand well before they can read a great deal of Rust code.

It's a well-written RFC, but I'm not sure the problem justifies the cost of the solution.

Aaron1011 commented:
To add on to @thomcc's point: I don't think a lint should be integrated into the type system this way.  I think an attribute similar to `#[must_use] ` would be a much better way of implementing this.

Nilstrieb commented:
I do like the idea of having a better way to deal with unintended implicit drops, but I think the proposed implementation has various flaws.
With this proposal, library authors have several things to do:
- On their public types, implement `Destruct` for all types that can be implicitly dropped (which will be almost all of them). This is a little annoying and it's easy to forget one, but an additional lint for this could help.
- On their public functions, they need to decide on how to handle `Destruct` bounds on their generics. This leaves them the following options:
  - Add `Destruct` bounds to most or all of their generics.
      - This is a breaking change and will force them to bump their major version.
      - Adding bounds restricts users to pass types that don't implement `Destruct` out of oversight or because the libraries of those types haven't been updated yet. Even though this is just a lint, it's impossible to `#[allow()]` passing the types anyways, even if they don't implement `Destruct` for the above reasons. People will have to write wrapper types that also implement the traits, which is quite annoying and can cause other incompatibilities.
  - Don't add any bounds and simply `allow()` the lint globally.
    - This causes their interface to be backwards compatible and as flexible as before.
    -  But this makes the RFC a bit useless, as people `allow()`ing the lint acts like as if it was never there before.
    - I think more libraries would decide on this, as it allows them to stay on their major version and doesn't restrict their interface too much.
- Many libraries in the ecosystem are unmaintained. These libraries will never have the possibitlity to implement `Destruct` for their types.

This already isn't great, but application authors are affected by this as well:
- Applications will need to make sure that all their own types (for those where it makes sense, which is most of them) implement `Destruct`. This is annoying to write and easy to forget (but forgetting it is not a problem as it will just cause the lint to fire, alerting the author that they forgot a type). If they use libraries, all those libraries will need to implement `Destruct` for their types. As mentioned above already, application authors also have no way to ignore bounds where they think that it can't cause problems. They will have to write wrapper types.

If we want to put this into the type system without polluting all code with it, it would need to be an auto trait and it would have to act like `Sized` on bounds. But this has many other disadvantages, so as @Aaron1011 pointed out already, I don't think the type system is the correct place for this.

chorman0773 commented:
> * Add Destruct bounds to most or all of their generics.
>    *  This is a breaking change and will force them to bump their major version.

And some libraries/applications have Minimum Support Rust Version requirements because of build requirements, so this is impossible for those libraries.

jashug commented:
Some counterarguments:

# Adding a `Destruct` bound is *not* a breaking change

Because `Destruct` is special cased so that failing a `T: Destruct` check produces a warning not an error, a library adding a `T: Destruct` bound at most produces warnings in that library's users. Another way of putting this is that for the purpose of type checking, every type implements `Destruct`, some of them are just noisy about it.

You can already see this behavior with the current implementation of `feature(const_trait_impl)`. Here is a Rust playground example where we can call a function with a `T: Destruct` bound on a type that does not explicitly implement `Destruct`. 
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2763dfa8f4e228c48f6d9f8c87722b48

Similarly, removing a `Destruct` implementation from a type is not a breaking change.

# `Destruct` should be an auto trait

... but the language team is opposed to adding new auto traits (see https://lang-team.rust-lang.org/design_notes/auto_traits.html). Acting like `Sized`, where it is assumed implicitly for all type variables, is even more frowned upon. Under those constraints, supporting `#[derive(Destruct)]` was the best I could do for a trait that almost every type will want to implement.

However, I am confused because the standard library (where `const Destruct` is already being used under `feature(const_trait_impl)`) does not seem to include any explicit declarations of `impl Destruct`. Looking at the code suggests that `const Destruct` may be using a custom structural algorithm, kind of like an auto trait but not? I don't know who to ask to get a better understanding of this situation.

If `Destruct` can be an auto trait or have a similar structural algorithm, then this RFC can be improved to remove `#[derive(Destruct)]` and add in a way of opting out, which would tremendously reduce the burden on users.

# Libraries will already be asked to add `~const Destruct` bounds to generic functions when `feature(const_trait_impl)` stabilizes

Look at the standard library source: `~const Destruct` bounds are not uncommon (for example, `Result::add`: https://doc.rust-lang.org/src/core/result.rs.html#1313-1317, or more stably click on `source` in https://doc.rust-lang.org/std/result/enum.Result.html#method.and).#3334 

# To be accurate, this lint needs most of the information present in the type system

In particular, I believe that when you are worried about accidentally dropping an important resource, it is very valuable to mark generic functions that can drop, and which types they can drop. The type/trait system is already set up to do this tracking and annotating, while a lint that cannot take advantage of the type system would be severely limited. Today, `must_use` is useful but too easy to accidentally bypass to be relied upon.

Nilstrieb commented:
> Because Destruct is special cased so that failing a T: Destruct check produces a warning not an error

Ah, I didn't notice that. I think that this is a really bad idea. Unsatisfied trait bounds always produce errors and never simply a warning, so this would be a huge special case. I think this nicely shows that the type system is not the correct place to put such a lint. This lint only fixes a, at the scale of the Rust language, rather small problem, so it should also only impact the Rust language in a small way. The current proposal is a massive change to the semantics of the Rust language, which is simply unwarranted for such a lint.

afetisov commented:
That would mean that the design of the lint is similar to `UnwindSafe` and `RefUnwindSafe`, and could be easily circumvented in similar ways. In fact, it isn't even that, because you propose violation of `T: Destruct` bound to not be a real violation at all, which is unprecedented in current Rust, and for a good reason. The design of those traits is regarded by many people as a mistake, and there is talk about entirely deprecating them. This proposal, being an even weaker lint with even smaller correctness benefits, is thus unlikely to fly.

Nilstrieb commented:
To elaborate on one of my points from yesterday, I think that these "bug prevention features" should have their usefulness and positive impact be directly proportional to their invasiveness and ways in which they make writing code more annoying.
The prime example of this is the borrow checker. The borrow checker is extremely invasive and one of the fundamental restrictions in Rust. But it's also extremely useful and catches many memory safety and logic errors all the time.
Another example is the `#[must_use]` attribute and its accompaning lint. Their invasiveness is rather small, a few types have it and some functions use it as well. It's also a useful lint that sometimes catches bugs, but nothing world-changing.

This RFC has a fairly small impact. It helps prevent a few bugs around `BufWriter` (and some more types out there, but not too many), which is useful but pretty limited. Therefore, its invasiveness should be equally small. But this RFC is very invasive, adding new special magic around "warn if not satisfied" bounds and expecting libraries to add these bounds in many places.

Ixrec commented:
I am also of the opinion that trying to make this work through generic code just isn't worth the complexity (especially since a hard guarantee / true linear typing is impossible anyway), and that an attribute similar to `#[must_use]` provides most of the potential benefit at a far lower cost.

`#[must_cleanup]` is my suggested strawman name for this design.

jashug commented:
The feedback here seems to be pretty universal that the type system is not the right place for this lint. I hear you, and rewrote the proposal to use an attribute. This relies on the compiler inferring the list of generic types that may be dropped, and propagating that to to crates that depend on this one.

With regards to the example: a warning should be produced on the `writer.write("Hello")?` line. When writing the example, I did not notice that the writer could be dropped there, and certainly didn't consciously decide to ignore errors on close. I changed the example to close and check for errors if the write fails instead. This is a perfect example of why this lint would be useful (though serious though would be needed to decide if `BufWriter` should make use of it).
